README.md
<div align="center">
<img src="https://github.com/IdeaLeap/Pipline/assets/49270362/eae58b53-77e9-41f2-913a-baeccdf4b962" style="width:100%;" alt="IdeaLeap Logo">
</div>
</p>
<p align="center">A Lightweight Low-code Pipeline code by GPT-4.</p>

## 📦 Installation

```bash
npm install @idealeap/pipeline
```

## ✨ Features

- 🔄 **Lightweight & Low-code**
- 🛠️ Pipeline Organization: Manage tasks easily
- 🔧 Custom Steps: Define preprocessing, post-processing, and more
- 🚄 Batch Support: Efficiently handle large data sets
- 📊 Event-Driven: Monitor progress seamlessly
- ⏳ Timeouts: Execute within limits
- ⚙️ Dependency Management: Ensure correct order
- 📈 Scalable: Add new step types easily

## 🚀 Usage

### Basic Usage

```typescript
const pipe1 = new Pipe((input, context) => input * 2, { id: "pipe1" });
const pipe2 = new Pipe((input, context) => input + 1, { id: "pipe2" });

const pipeline = new Pipeline([pipe1, pipe2]);
pipeline.execute(2).then((result) => console.log(result));
```

### Event-Driven

```typescript
const pipe1 = new Pipe<number, number>(
  (input) => {
    return input + 1;
  },
  {
    id: "pipe1",
  },
);

const pipe2 = new Pipe<number, number>(
  (input) => {
    return input * 2;
  },
  {
    id: "pipe2",
  },
);

const pipeline = new Pipeline([pipe1, pipe2], {
  onProgress: (completed, total) => {
    console.log(`Progress: ${completed}/${total}`);
  },
});

// 执行管道
await pipeline.execute(1).then((results) => {
  console.log("Final results:", results);
});
```

### Batch Support & Dynamic JS Executor

```typescript
const pipe1 = new Pipe(
  (input: string) => {
    return input + "——————(被我测了";
  },
  {
    id: "pipe1",
    batch: true,
    onBatchResult: (x: string[]) => {
      return `*${x.join("\n")}*`;
    },
  },
);
const pipe2 = new Pipe(
  async (input: string) => {
    await new Promise((resolve) => setTimeout(resolve, 1000));
    return input + "\n\n你看看你测试了多少！！🤬😡";
  },
  {
    id: "pipe2",
  },
);
const pipe3 = new Pipe(
  async (input: string) => {
    const res = await DynamicExecutor.run({
      code: `console.log(\`${input}\`);
      return \`${input}\`;`,
    });
    console.log(res);
    return input;
  },
  {
    id: "pipe3",
  },
);
const pipeline = new Pipeline([pipe1, pipe2, pipe3], {
  onProgress: (completed, total) => {
    console.log(`Progress: ${completed}/${total}`);
  },
});

// 执行管道
const res = await pipeline.execute([
  "我是甲，别测我┭┮﹏┭┮",
  "我是乙，求你测我┭┮﹏┭┮",
  "我是丙，来啊你以为我怕你！",
  "我是丁，你敢？！滚一边去~",
]);
```

### Chain Call

```typescript
const pipeline = Pipeline.create()
  .addPipe(
    Pipe.create((input: number) => input + 1, {
      id: "step1",
    }).setDescription("Increment by 1"),
  )
  .addPipe(
    Pipe.create((input: number) => input * 2, {
      id: "step2",
    }).setDescription("Multiply by 2"),
  )
  .setOnProgress((completed, total) => {
    console.log(`Progress: ${completed}/${total}`);
  });

// 执行
await pipeline.execute(1).then((result) => {
  console.log("Final result:", result);
});

// 序列化为 JSON
const jsonConfig = JSON.stringify(pipeline.toJSON());
console.log("Serialized config:", jsonConfig);
```

### Function Map

```typescript
// 示例
const jsonConfig: SerializablePipelineOptions = {
  pipes: [{ id: "step1" }, { id: "step2", timeout: 1000 }],
};

const fnMap = {
  step1: (input: string) => `${input}-step1`,
  step2: (input: string) => `${input}-step2`,
};

const pipeline = Pipeline.fromJSON(jsonConfig, fnMap);

// 执行 Pipeline
await pipeline.execute("我饿").then(console.log);
```

### Low Code

```typescript
const pipeRegistry = PipeRegistry.init();
// 注册预定义的 Pipe 类型
pipeRegistry.register("FetchData", async () => {
  // 这里用一个简单的 setTimeout 来模拟异步数据获取
  return new Promise((resolve) =>
    setTimeout(() => resolve("fetched data"), 1000),
  );
});

pipeRegistry.register("TransformData", () => {
  // 这里只是简单地返回一个字符串，实际情况可能涉及到更复杂的数据转换
  // console.log(input, context);
  return "transformed data";
});

pipeRegistry.register("postProcess", (input: string) => {
  // 这里只是简单地返回一个字符串，实际情况可能涉及到更复杂的数据转换
  // console.log(input, context);
  return input + "\nBy the way, I'm postProcess";
});

const pipelineJson = {
  pipes: [
    {
      id: "FetchData",
      use: "FetchData",
    },
    {
      id: "TransformData",
      use: "TransformData",
      postProcessUse: "postProcess",
    },
  ],
};

const pipeline = Pipeline.fromJSON(pipelineJson, {}, pipeRegistry);

// 序列化为 JSON
const jsonConfig = JSON.stringify(pipeline.toJSON());
console.log("Serialized config:", jsonConfig);

// 导入 JSON
const newPipeline = Pipeline.fromJSON(JSON.parse(jsonConfig), {}, pipeRegistry);

// 执行
await newPipeline.execute().then(console.log);
```

## 📄 Author

(C) 2023 GPT-4, Marlene, Idealeap


CHANGELOG.md
# Changelog

## [1.2.0](https://github.com/IdeaLeap/Pipeline/compare/v1.1.0...v1.2.0) (2023-10-11)


### Features

* add DynamicExecutor Fn ([c916fc8](https://github.com/IdeaLeap/Pipeline/commit/c916fc860b6574eb47fb168464ac3ea76fd8940d))
* V1.1.6 ([336a828](https://github.com/IdeaLeap/Pipeline/commit/336a828debcdff48252303b2255c6ed223fd872b))
* 修改type为use ([3323909](https://github.com/IdeaLeap/Pipeline/commit/3323909fbeb517f090a67580cde28763c1fbbc82))
* 支持API文档自动生成 ([c87e72a](https://github.com/IdeaLeap/Pipeline/commit/c87e72abed00bd29c25e2749dec22c833c147ec7))
* 支持build直接publish ([37de50d](https://github.com/IdeaLeap/Pipeline/commit/37de50d6a897cf0fa6cafa650dfce12c59150466))
* 支持globalParams ([f25abfa](https://github.com/IdeaLeap/Pipeline/commit/f25abfa474bb5042c5873d993de062f6b089d98e))
* 支持pipe的input指定 && 更强的插槽渲染 ([04bfd9d](https://github.com/IdeaLeap/Pipeline/commit/04bfd9d4dd0d0ac55df725d460970d7ec7528ddf))
* 支持pipe的input指定 && 更强的插槽渲染 ([bb0e617](https://github.com/IdeaLeap/Pipeline/commit/bb0e617d663fd71362d1b6de14ed49abc9bfc66f))
* 支持tsc编译 ([9cbd128](https://github.com/IdeaLeap/Pipeline/commit/9cbd12878267446d6956d9b67aecff5f8ba8712d))
* 支持插槽渲染 && 使用lodash优化代码 ([ef91a16](https://github.com/IdeaLeap/Pipeline/commit/ef91a165efbf3ae8045ff971e6677ec37bdd0a9f))
* 支持获取self_params ([c8a6447](https://github.com/IdeaLeap/Pipeline/commit/c8a6447807f6e95783baec558b6cc6bb9a713382))


### Bug Fixes

* emitter err ([05b634f](https://github.com/IdeaLeap/Pipeline/commit/05b634f098d88b86c033111285519907e6347125))
* publish ([9af58ca](https://github.com/IdeaLeap/Pipeline/commit/9af58cab14099bc168fad65111c99ec1f0d9f520))
* 增加深拷贝选项 ([01f58b8](https://github.com/IdeaLeap/Pipeline/commit/01f58b80c3868902be069a69573a318ddaa94c4e))
* 开发环境优化 ([b1db2cd](https://github.com/IdeaLeap/Pipeline/commit/b1db2cdd36ea17304d6791a434270eb339a8a4e7))
* 禁止设置已用ID ([6dd96e8](https://github.com/IdeaLeap/Pipeline/commit/6dd96e886da233739d7e973d79946d40c0fe10b3))
* 部分错误 ([6a232f5](https://github.com/IdeaLeap/Pipeline/commit/6a232f5525dd439f85f7ca67b6e9e94989c740fb))
* 错误输入导致的ts问题 ([d2a57b3](https://github.com/IdeaLeap/Pipeline/commit/d2a57b3db94c4881cb8ec722159c1471de85a491))

## [1.1.0](https://github.com/IdeaLeap/Pipline/compare/v1.0.0...v1.1.0) (2023-09-02)

### Features

- auto deploy ([33eb5cc](https://github.com/IdeaLeap/Pipline/commit/33eb5cc8763b49dcbe2bfb1072a7b24c78af78d2))

### Bug Fixes

- pnpm ([a21a2ce](https://github.com/IdeaLeap/Pipline/commit/a21a2cee16bae2e32652302445addc51660442f3))
- pnpm ([e5bd5c5](https://github.com/IdeaLeap/Pipline/commit/e5bd5c51c9d666eaa4c319c265c640115071df7d))
- pnpm lock ([39719d6](https://github.com/IdeaLeap/Pipline/commit/39719d6140e4023c27d758c2168e8200ac1bc220))

## 1.0.0 (2023-09-02)

### Features

- 支持预置参数 ([7e67126](https://github.com/IdeaLeap/Pipline/commit/7e6712637a8faaa2f90249f447ba457456866381))
- 显式注册函数插桩支持 ([a22db59](https://github.com/IdeaLeap/Pipline/commit/a22db5963e7ddceac05680548838bcfece65357d))
- 更好的低代码支持 ([664163f](https://github.com/IdeaLeap/Pipline/commit/664163fc92ea29499cfbc5372c80dab4d038e1cd))


package/__tests__/pipe.test.ts
import {
  Pipe,
  Pipeline,
  SerializablePipelineOptions,
  PipeRegistry,
  DynamicExecutor,
  EventEmitter,
  PipeOptions,
  PipelineOptions,
  PipelineContext,
} from "@idealeap/pipeline"; // 请替换成你的模块导入方式

test("Pipeline", async () => {
  const pipe1 = new Pipe<number, number>(
    (input) => {
      return input + 1;
    },
    {
      id: "pipe1",
    },
  );

  const pipe2 = new Pipe<number, number>(
    (input) => {
      return input * 2;
    },
    {
      id: "pipe2",
    },
  );

  const pipeline = new Pipeline([pipe1, pipe2], {
    onProgress: (completed, total) => {
      console.log(`Progress: ${completed}/${total}`);
    },
  });

  // 执行管道
  await pipeline.execute(1).then((results) => {
    console.log("Final results:", results);
  });
});

test("Pipeline with 并行&串行&动态代码", async () => {
  const pipe1 = new Pipe(
    (input: string) => {
      return input + "——————(被我测了";
    },
    {
      id: "pipe1",
      batch: true,
      onBatchResult: (x: string[]) => {
        return `*${x.join("\n")}*`;
      },
    },
  );
  const pipe2 = new Pipe(
    async (input: string) => {
      await new Promise((resolve) => setTimeout(resolve, 1000));
      return input + "\n\n你看看你测试了多少！！🤬😡";
    },
    {
      id: "pipe2",
    },
  );
  const pipe3 = new Pipe(
    async (input: string) => {
      const res = await DynamicExecutor.run({
        code: `console.log(\`${input}\`);
        return \`${input}\`;`,
      });
      console.log(res);
      return input;
    },
    {
      id: "pipe3",
    },
  );
  const pipeline = new Pipeline([pipe1, pipe2, pipe3], {
    onProgress: (completed, total) => {
      console.log(`Progress: ${completed}/${total}`);
    },
  });

  // 执行管道
  const res = await pipeline.execute([
    "我是甲，别测我┭┮﹏┭┮",
    "我是乙，求你测我┭┮﹏┭┮",
    "我是丙，来啊你以为我怕你！",
    "我是丁，你敢？！滚一边去~",
  ]);
  expect([...res.values()].at(-1)).toEqual(
    `*我是甲，别测我┭┮﹏┭┮——————(被我测了\n我是乙，求你测我┭┮﹏┭┮——————(被我测了\n我是丙，来啊你以为我怕你！——————(被我测了\n我是丁，你敢？！滚一边去~——————(被我测了*\n\n你看看你测试了多少！！🤬😡`,
  );
});

test("Pipeline with 链式调用", async () => {
  const pipeline = Pipeline.create()
    .addPipe(
      Pipe.create((input: number) => input + 1, {
        id: "step1",
      }).setDescription("Increment by 1"),
    )
    .addPipe(
      Pipe.create((input: number) => input * 2, {
        id: "step2",
      }).setDescription("Multiply by 2"),
    )
    .setOnProgress((completed, total) => {
      console.log(`Progress: ${completed}/${total}`);
    });

  // 执行
  await pipeline.execute(1).then((result) => {
    console.log("Final result:", result);
  });

  // 序列化为 JSON
  const jsonConfig = JSON.stringify(pipeline.toJSON());
  console.log("Serialized config:", jsonConfig);
});

test("Pipeline with JSON", async () => {
  // 示例
  const jsonConfig: SerializablePipelineOptions = {
    pipes: [{ id: "step1" }, { id: "step2", timeout: 1000 }],
  };

  const fnMap = {
    step1: (input: string) => `${input}-step1`,
    step2: (input: string) => `${input}-step2`,
  };

  const pipeline = Pipeline.fromJSON(jsonConfig, fnMap);

  // 执行 Pipeline
  await pipeline.execute("我饿").then(console.log);
});

test("Pipeline with pipeRegistry", async () => {
  const pipeRegistry = PipeRegistry.init();
  // 注册预定义的 Pipe 类型
  pipeRegistry.register("FetchData", async () => {
    // 这里用一个简单的 setTimeout 来模拟异步数据获取
    return new Promise((resolve) =>
      setTimeout(() => resolve("fetched data"), 1000),
    );
  });

  pipeRegistry.register("TransformData", () => {
    // 这里只是简单地返回一个字符串，实际情况可能涉及到更复杂的数据转换
    // console.log(input, context);
    return "transformed data";
  });

  pipeRegistry.register("postProcess", (input: string) => {
    // 这里只是简单地返回一个字符串，实际情况可能涉及到更复杂的数据转换
    // console.log(input, context);
    return input + "\nBy the way, I'm postProcess";
  });

  const pipelineJson = {
    pipes: [
      {
        id: "FetchData",
        use: "FetchData",
      },
      {
        id: "TransformData",
        use: "TransformData",
        postProcessUse: "postProcess",
      },
    ],
  };

  const pipeline = Pipeline.fromJSON(pipelineJson, {}, pipeRegistry);

  // 序列化为 JSON
  const jsonConfig = JSON.stringify(pipeline.toJSON());
  console.log("Serialized config:", jsonConfig);

  // 导入 JSON
  const newPipeline = Pipeline.fromJSON(
    JSON.parse(jsonConfig),
    {},
    pipeRegistry,
  );

  // 执行
  await newPipeline.execute().then(console.log);
  expect(jsonConfig).toEqual(
    `{"pipes":[{"id":"FetchData","use":"FetchData"},{"id":"TransformData","use":"TransformData","postProcessUse":"postProcess"}]}`,
  );
});
test("Pipe的Emitter和依赖", async () => {
  // 创建Pipe实例
  const pipe1Options: PipeOptions<number, number> = { id: "pipe1" };
  const pipe1 = new Pipe<number, number>((input) => {
    return input * 2;
  }, pipe1Options);

  const pipe2Options: PipeOptions<number, string> = {
    id: "pipe2",
  };
  const pipe2 = new Pipe<number, string>((input, context) => {
    const depResult = context.stepResults["pipe1"];
    return `${input + depResult}`;
  }, pipe2Options);

  // 创建全局EventEmitter实例
  const globalEmitter = new EventEmitter();

  globalEmitter.on(
    "stepComplete",
    (step: any, totalSteps: any, result: any) => {
      console.log(
        `Step ${step}/${totalSteps} completed with result: ${result}`,
      );
    },
  );

  globalEmitter.on("error", (error) => {
    console.log(`An error occurred: ${error}`);
  });

  // 创建ChainPipeOptions
  const chainPipeOptions: PipelineOptions = {
    onProgress: (completed, total) => {
      console.log(`Progress: ${completed}/${total}`);
    },
    emitter: globalEmitter,
  };

  // 执行管道
  async function run() {
    try {
      const _ = new Pipeline([pipe1, pipe2], chainPipeOptions);
      const result = await _.execute(2);
      console.log("Final result:", result);
    } catch (error) {
      console.log("Error:", error);
    }
  }

  await run();
});

test("Pipe获取初始输入和当前参数", async () => {
  const pipeRegistry = PipeRegistry.init();
  pipeRegistry.register(
    "step1",
    async (input: any, context: PipelineContext) => {
      console.log("step1", input, context.stepParams["self_params"]);
      return new Promise((resolve) =>
        setTimeout(() => resolve(input + "🚺"), 1000),
      );
    },
  );

  pipeRegistry.register("step2", (input: any, context: PipelineContext) => {
    console.log(
      "step2",
      input,
      context.stepParams["self_params"],
      context.stepResults["index_input"],
    );
    return context.stepResults["index_input"];
  });

  pipeRegistry.register(
    "step3",
    async (input: any, context: PipelineContext) => {
      console.log("step3", input, context.stepParams["self_params"]);
      return new Promise((resolve) => setTimeout(() => resolve(input), 1000));
    },
  );

  const pipelineJson = {
    pipes: [
      {
        id: "step1_",
        use: "step1",
        params: { test: "test!!" },
      },
      {
        id: "step2_",
        use: "step2",
        inputs: {
          FetchData: "step1_",
        },
        params: { test: "test22!!{{FetchData}}" }, //插槽！
      },
      {
        id: "step3_",
        use: "step3",
        inputs: {
          input: "step1_", //这里的input是step1_的结果
        },
      },
    ],
  };

  const pipeline = Pipeline.fromJSON(pipelineJson, {}, pipeRegistry);

  await pipeline.execute("我是输入").then(console.log);
});

test("Pipe使用globalParams", async () => {
  const pipeRegistry = PipeRegistry.init();
  pipeRegistry.register(
    "step1",
    async (input: any, context: PipelineContext) => {
      console.log(input, context.stepParams["self_params"]);
      return new Promise((resolve) => setTimeout(() => resolve(input), 1000));
    },
  );
  const pipelineJson = {
    pipes: [
      {
        id: "FetchData",
        use: "step1",
        params: { test: "test!!" },
      },
    ],
    globalParams: {
      abc: "123",
    },
  };

  const pipeline = Pipeline.fromJSON(pipelineJson, {}, pipeRegistry);

  await pipeline.execute("我是输入").then(console.log);
});


package/__tests__/utils.test.ts
import { replaceSlots } from "@idealeap/pipeline";

test("replaceSlots", () => {
  const testObj = {
    text: "Hello {{name}}",
    info: {
      age: "{{age}} years old",
      address: "{{addressObj.city}}124453",
      others: "{{addressObj}}",
    },
  };

  const replacements = {
    name: "Alice",
    age: "25",
    addressObj: {
      street: "123 Main St",
      city: "Example City",
    },
  };

  console.log(replaceSlots(testObj, replacements));
  expect(replaceSlots(testObj, replacements)).toEqual({
    text: "Hello Alice",
    info: {
      age: "25 years old",
      address: "Example City124453",
      others: { street: "123 Main St", city: "Example City" },
    },
  });
  debugger;
});


package/__tests__/registry.test.ts
import { PipeRegistry } from "@idealeap/pipeline";

test("PipeRegistry", () => {
  const pipeRegistry = PipeRegistry.init();
  console.log(pipeRegistry);
  debugger;
});


package/registry.ts
import { MaybePromise, PipelineContext } from "@idealeap/pipeline/pipe";
import { DynamicExecutor } from "@idealeap/pipeline/executor/index";
import lodash from "lodash";
export type PipeRegistryType = PipeRegistry;

export type Registry = Record<
  string,
  (input: any, context: PipelineContext) => MaybePromise<any>
>;

export class PipeRegistry {
  private registry: Registry = {};

  register(
    type: string,
    callback: (input: any, context: PipelineContext) => MaybePromise<any>,
  ) {
    this.registry[type] = callback;
  }

  get(type: string) {
    return this.registry[type];
  }

  static commonPreProcess = {
    validateData: (input: any, context: PipelineContext) => {
      // 数据验证逻辑
      console.log("validateData", input, context);
      return input;
    },
  };

  static commonPostProcess = {
    logData: (result: any, context: PipelineContext) => {
      console.log(result, context);
      return result;
    },
  };

  static customFn: Record<
    string,
    (input: any, context: PipelineContext) => any
  > = {};

  _Fn = {
    DynamicExecutor: async (input: any, context: PipelineContext) => {
      const params = context.stepParams["self_params"];
      const executor = new DynamicExecutor(params);
      if (lodash.isObject(params.input)) {
        return await executor.execute<any>(params.code, ...params.input);
      } else {
        return await executor.execute<any>(params.code, params.input);
      }
    },
  };

  static init() {
    const pipeRegistry = new PipeRegistry();
    //遍历commonPreProcess进行pipeRegistry.register注册
    Object.entries(PipeRegistry.commonPreProcess).forEach(([type, fn]) => {
      pipeRegistry.register(type, fn);
    });
    //遍历commonPostProcess进行pipeRegistry.register注册
    Object.entries(PipeRegistry.commonPostProcess).forEach(([type, fn]) => {
      pipeRegistry.register(type, fn);
    });
    //遍历customFn进行pipeRegistry.register注册
    Object.entries(PipeRegistry.customFn).forEach(([type, fn]) => {
      pipeRegistry.register(type, fn);
    });
    Object.entries(pipeRegistry._Fn).forEach(([type, fn]) => {
      pipeRegistry.register(type, fn);
    });
    return pipeRegistry;
  }
}


package/batch/batch.ts
// 本代码由GPT4生成，具体可见https://pandora.idealeap.cn/share/26cf4f3d-9089-4dbc-87ef-63bc01ed5bb2
export type Awaited<T> = T extends PromiseLike<infer U> ? U : T;
// 批处理选项
export interface BatchOptions<T, R> {
  onData?: (input: T) => T;
  onBatchData?: (inputs: T[]) => T[] | T;
  onResult?: (result: Awaited<R>) => Awaited<R>;
  onBatchResult?: (results: Awaited<R>[]) => Awaited<R>[] | Awaited<R>;
  onProgress?: (completed: number, total: number) => void;
  onError?: (error: any) => void;
}

export function batchDecorator<T, R>(
  fn: ((input: T) => Promise<R> | R) | ((input: T) => Promise<R> | R)[],
  options?: BatchOptions<T, R>,
): (input: T | T[]) => Promise<Awaited<R> | Awaited<R>[]> {
  return async (input: T | T[]): Promise<Awaited<R> | Awaited<R>[]> => {
    const { onBatchData, onBatchResult } = options || {};
    const total = Array.isArray(input) ? input.length : 1;
    const isMultipleFns = Array.isArray(fn);
    const functions = isMultipleFns
      ? (fn as ((input: T) => Promise<R> | R)[])
      : [fn as (input: T) => Promise<R> | R];

    let inputs = Array.isArray(input) ? input : [input];

    if (onBatchData) {
      const processedBatch = onBatchData(inputs);
      inputs = Array.isArray(processedBatch)
        ? processedBatch
        : Array(total).fill(processedBatch);
    }

    let allResults: Awaited<R>[] = [];
    let completedCount = 0;
    async function runSingle<T, R>(
      data: T,
      f: (input: T) => Promise<R> | R,
      options: BatchOptions<T, R>,
    ): Promise<Awaited<R> | null> {
      const { onData, onResult, onProgress, onError } = options;
      try {
        const processedData = onData ? onData(data) : data;
        const result = await Promise.resolve(f(processedData));
        const finalResult = onResult ? onResult(result as Awaited<R>) : result; // 类型断言
        onProgress?.(++completedCount, total * functions.length);
        return finalResult as Awaited<R>; // 类型断言
      } catch (error) {
        onError?.(error);
        return null;
      }
    }

    for (const singleFn of functions) {
      const results: Promise<Awaited<R> | null>[] = [];
      for (const data of inputs) {
        results.push(runSingle(data, singleFn, options || {}));
      }
      const resultBatch = await Promise.all(results);
      allResults = allResults.concat(
        resultBatch.filter(Boolean) as Awaited<R>[],
      );
    }

    if (onBatchResult) {
      const finalBatchResult = onBatchResult(allResults);
      return finalBatchResult;
    }

    return allResults;
  };
}


package/executor/executor.ts
// 本代码由GPT4生成，具体可见https://pandora.idealeap.cn/share/07ee53fc-e074-46f6-b808-b4f159bb3468
export interface ExecutorConfig {
  timeout?: number;
  logging?: boolean;
  environment?: Record<string, any>;
  allowedBuiltins?: string[];
  beforeExecute?: (code: string) => void;
  afterExecute?: (result: any, code: string, error?: Error) => void;
  validateCode?: (code: string) => boolean;
}

export class DynamicExecutor {
  private config: ExecutorConfig;

  constructor(config?: ExecutorConfig) {
    this.config = config || {};
  }

  private log(message: string) {
    if (this.config.logging) {
      console.log(`[DynamicExecutor]: ${message}`);
    }
  }

  public async execute<T = any>(
    code: string,
    ...args: any[]
  ): Promise<T | null> {
    let result: T | null = null;
    let executionError: Error | undefined = undefined;

    if (
      typeof code !== "string" ||
      (this.config.validateCode && !this.config.validateCode(code))
    ) {
      throw new Error("Invalid or unsafe code");
    }

    // Call the beforeExecute hook if provided
    this.config.beforeExecute?.(code);

    this.log(`Executing code: ${code}`);

    try {
      const sandbox: Record<string, any> = {
        // fetch: this.config.allowedBuiltins?.includes("fetch")
        //   ? fetch
        //   : undefined,
        ...this.config.environment,
      };
      sandbox["args"] = args;

      const envKeys = Object.keys(sandbox).join(",");
      const envValues = Object.values(sandbox);

      // eslint-disable-next-line @typescript-eslint/no-implied-eval
      const func = new Function(
        envKeys,
        `
        "use strict";
        return (async () => {
          ${code};
        })();
      `,
      ).bind(null, ...envValues);

      if (this.config.timeout) {
        const promise = func();
        result = (await Promise.race([
          promise,
          new Promise<T | null>((_, reject) =>
            setTimeout(() => reject(new Error("Timeout")), this.config.timeout),
          ),
        ])) as T;
      } else {
        result = (await func()) as T;
      }
    } catch (error: unknown) {
      executionError = error as Error;
      this.log(`Error occurred: ${String(executionError)}`);
      result = null;
      throw executionError;
    }

    // Call the afterExecute hook if provided
    this.config.afterExecute?.(result, code, executionError);

    this.log(`Execution result: ${String(result)}`);
    return result;
  }

  static async run(params: { code: string; config?: ExecutorConfig }) {
    const { code, config, ...rest } = params;
    const executor = new DynamicExecutor(config);
    const res = await executor.execute(code, ...Object.values(rest));
    return res;
  }
}


package/utils.ts
import lodash from "lodash";

/**
 * Replaces all slots in the given object with the provided replacements.
 * If a slot matches a key in the replacements, it will be substituted.
 * If the replacement value is an object, and the slot covers the entire string,
 * the entire value will be replaced by the object. Otherwise, the object will be stringified.
 *
 * @param obj - The object containing slots to be replaced.
 * @param replacements - The replacements to use for the slot substitution.
 * @returns - A new object with all the slots replaced.
 *
 * @example
 * ```typescript
 * const testObj = {
 *   text: "Hello {{name}}",
 *   info: {
 *     age: "{{age}} years old",
 *     address: "{{addressObj.city}}124453"
 *   }
 * };
 *
 * const replacements = {
 *   name: "Alice",
 *   age: "25",
 *   addressObj: {
 *     street: "123 Main St",
 *     city: "Example City"
 *   }
 * };
 *
 * console.log(replaceSlots(testObj, replacements));
 * ```
 */
export function replaceSlots(
  obj: Record<string, any>,
  replacements: Record<string, any>,
): any {
  return lodash.cloneDeepWith(obj, (value: any) => {
    if (lodash.isString(value)) {
      // If the whole value is a slot
      const fullMatchReplacement = /^{{\s*(.*?)\s*}}$/.exec(value);
      if (fullMatchReplacement && !!fullMatchReplacement[1]) {
        const replacementValue = lodash.get(
          replacements,
          fullMatchReplacement[1],
        );
        if (replacementValue !== undefined) {
          return replacementValue;
        }
      }

      // If part of the value is a slot
      return value.replace(/{{\s*(.*?)\s*}}/g, (match, path) => {
        const replacementValue = lodash.get(replacements, path);
        return replacementValue !== undefined
          ? String(replacementValue)
          : match;
      });
    }
  });
}

/**
 * Merges two JSON objects safely by checking for key conflicts.
 * If a key exists in both `obj1` and `obj2`, it will throw an error.
 *
 * @param obj1 - The first object to merge.
 * @param obj2 - The second object to merge.
 * @returns A merged object containing properties from both `obj1` and `obj2`.
 * @throws {Error} Throws an error if a key exists in both `obj1` and `obj2`.
 *
 * @example
 * const objA = { a: 1, b: 2 };
 * const objB = { c: 3 };
 * mergeJSONSafely(objA, objB); // Returns { a: 1, b: 2, c: 3 }
 */
export function mergeJSONSafely(obj1: object, obj2: object): object {
  lodash.mergeWith(obj1, obj2, (objValue, srcValue, key, object, source) => {
    if (lodash.has(obj1, key) && lodash.has(obj2, key)) {
      throw new Error(`Pipe Params ${key} 与 globalParams 冲突`);
    }
    return undefined; // 返回undefined以使用默认的合并行为
  });
  return obj1;
}

/**
 * Determines if a variable is a plain object with string values.
 *
 * @param variable - The variable to check.
 * @returns `true` if the variable is a plain object and all its values are strings, otherwise returns `false`.
 *
 * @example
 * isRecordOfString({ a: 'hello', b: 'world' }); // Returns true
 * isRecordOfString({ a: 'hello', b: 123 });     // Returns false
 */
export function isRecordOfString(
  variable: any,
): variable is Record<string, string> {
  return (
    lodash.isPlainObject(variable) && lodash.every(variable, lodash.isString)
  );
}
